Re-do user tag voting UI to address user complaints

bdach opened the pull request:
https://github.com/user-attachments/assets/c457fd75-ad3a-4706-8a50-975fdfa27359

Main points this intends to hit:

- The popover is replaced by a drawer type view due to complaints that the popover was too small, and that adding multiple tags in one go was cumbersome because the popover closed every time. (The second part may seem like a weak reason to do this - after all, you could make the popover *not* close - however, the hidden reason to do this is that the popover was attached to the plus button, which was under the tag flow, which meant as you add more tags, the flow can shift downwards, and therefore so can the popover, all of which sucks)
- Better organisation by grouping & sorting alphabetically
- The tags no longer move from the right list to the left - the right list always has the full list of tags (avoids confusion of "where is XYZ tag that I'm looking for"). The left list shows most popular tags, and the right list shows everything.
- Voting is possible from both lists, and both lists show the voted status.

---

- Intended to address https://github.com/ppy/osu/discussions/32568#discussioncomment-12612884
- Closes https://github.com/ppy/osu/issues/32630, I guess
- [x] Depends on https://github.com/ppy/osu/pull/32657

bdach committed:
Redesign vote buttons to better work with hierarchical tags

bdach committed:
Move the extra tag selector out of popover

bdach committed:
Adjust transform

bdach committed:
Add groupings to all tags list

bdach committed:
Adjust flow direction

bdach committed:
Always show all tags in full list & allow voting from both views

bdach committed:
Final sizing adjustments

bdach committed:
Async load panels to avoid update thread hitching

bdach committed:
Fix enter-to-select-tag interaction

cyperdark commented:
i dont think we need to have group name moved above all subcategories, and i prefer having category in title like this: category/subcategory. That way it easier to see in which category you in while scrolling, and it eliminate name overlaps with other categories (if there any)


![image](https://github.com/user-attachments/assets/77259db7-b1f1-4196-b208-ee1ec5bd7c41)

minetoblend commented on osu.Game/Screens/Ranking/UserTagControl.cs:
```diff
@@ -276,8 +325,8 @@ public DrawableUserTag(UserTag userTag)
-                Origin = Anchor.Centre;
+                Anchor = Anchor.CentreLeft;
+                Origin = Anchor.CentreLeft;
                 CornerRadius = 8;
                 Masking = true;
```
I somehow missed that the masking is needed for the glow effect, however the mismatch inn corner radius makes the glow look a bit weird, almost like it's disappearing on the corners

![image](https://github.com/user-attachments/assets/18bd3a7d-b466-4a53-8245-a9cd3b662072)

If both radii match it looks normal again

![image](https://github.com/user-attachments/assets/ff4d948d-989b-4cfc-97ac-135a565d18ed)

peppy commented:
Just some initial thoughts:

- The "thumbs up" not matching the visual display of voted-for items on the left is weird and arbitrary.
- The sizing of tags on the left looks very disorganised. Are they supposed to be both-direction flowing but are shown one-per-line due to their size?
- The area at the right always being visible feels bad to me (for the general case of users not interacting with it, it feels very visually busy). Can we move it back to a popover but have it not dismiss until the user is finished working with it or something?

bdach commented:
@minetoblend I'll apply the simplest of your fixes like corner radius or font weight and such, but stuff like animation polish and such is out of scope IMO because this is already a large diff and can be done separately later. Especially so given @peppy's comment that throws doubt on most of the direction here (which I, to be quite honest, half-expected beforehand, because UX is deeply subjective and even after however many years that I've been doing things here, I may have like a 70% hit rate when I invent designs, which is still not 100% and I'm not sure that it will ever be).

---

@peppy:

> The "thumbs up" not matching the visual display of voted-for items on the left is weird and arbitrary.

What would you expect to change? For them to be unified? In what way, to be more like the left or the right?

I'm not showing vote counts on the right, because there's not really a guarantee I have a *correct* vote count to show on the right. [web limits top tags returned to top 50](https://github.com/ppy/osu-web/blob/e5e98872312f2f3098d08ad03653bad0e1f4983c/app/Models/Beatmap.php#L408), which means I might not even know how many votes a tag has if the tail is sufficiently long. If you think that's unimportant enough to disregard, then I can do that.

> The sizing of tags on the left looks very disorganised. Are they supposed to be both-direction flowing but are shown one-per-line due to their size?

They were both-direction flowing but basically ended up being one-per-line anyway and therefore are now vertical-flowing.

I don't know how to address the "disorganisation" concern, I'd say what we have on master is already disorganised, especially when there's like 10-15 tags applied.

> The area at the right always being visible feels bad to me (for the general case of users not interacting with it, it feels very visually busy). Can we move it back to a popover but have it not dismiss until the user is finished working with it or something?

I touched on this in the OP, and to @frenzibyte when he suggested in on discord before: where do you attach a popover to so that it doesn't start moving all over when you start tagging? Do you think a popover is suitable given that based on feedback it seems that most users want to assign double digit number of tags every time they interact with this control?

I'm not one to want to pull the "crowd opinion" move, but note that the reaction to the OP is significantly positive. There's also [specific feedback about not wanting the popover to be a popover](https://github.com/ppy/osu/discussions/32568#discussioncomment-12613280).

I guess end of day the buck stops with you but I don't really think going back to a popover is going to address user needs here.

peppy commented:
Further discussion in voice leading to some proposals:

- Always show thumbs up button (use same colour differentiation for "vote state" as on the left)
- Have the right hand interface only visible after opening it with a button (like a drawer)
- Change the tags to flow in all directions again (so they flow in the full available area when the drawer isn't open)

minetoblend commented:
> but stuff like animation polish and such is out of scope IMO because this is already a large diff and can be done separately later

Alright, should I post those things again as follow-up prs once this is merged?

bdach committed:
Unify voted displays

bdach committed:
Compromise between popover and persistent view with a slideout

bdach committed:
Fix corner radii not matching

bdach committed:
Adjust font weights

bdach committed:
Adjust glow

peppy commented:
I think this looks much better.

bdach committed:
Merge branch 'ruleset-specific-user-tags' into better-user-tag-ui

bdach commented:
Heads up that this depends on https://github.com/ppy/osu/pull/32657 now.

bdach committed:
Merge branch 'master' into better-user-tag-ui

peppy committed:
Adjust mouse down animation in a better way (and apply to small tag buttons too)

peppy committed:
Remove flow animations

They don't look great, so let's just get something out which doesn't
involve things flying everywhere to start with.

peppy committed:
Switch to using a popover again

peppy commented:
Had a play with taking this new UI and putting it [back in a popover](https://github.com/peppy/osu/commit/23532eac55adc07726d37be547bc79687660c4ca) again (while hopefully not regressing any of the concerns):

https://github.com/user-attachments/assets/70cdf48b-ce63-4d2b-bc8c-27a8cfaff63e

Will need to fix up the loading layer (another weirdness I'd like to remove, so that works) to hopefully be back in per-tag displays and allowing synchronous requests to fire, but I think this could work?

I'll take a look at fixing the loading part tomorrow unless you want to grab it.

minetoblend commented:
The "add" button being on the left is very unconventional for this type of component, I think placing it at the end (inline or with a line-break) would be a better position for it.

bdach commented:
> I'll take a look at fixing the loading part tomorrow unless you want to grab it.

I kinda don't if I'm honest. I'd rather go into web and get things into a better shape there so that the complexity of implementation could be offloaded off client (and also to decrease the chance of users 429ing themselves due to spamming the control).

> The "add" button being on the left is very unconventional for this type of component, I think placing it at the end (inline or with a line-break) would be a better position for it.

If you put it at the end it'll start moving as you add more tags. Which is why I wanted to have this view either static or as a drawer, but I've been overruled so :shrug:

peppy commented:
The main concern I have is:

- One operation at a time will lead to a frustrating experience when assigning tags quickly (like i'm used to doing on github, for instance)
- The loading layer thing looks a touch not just in this component but anywhere really (and if we switch to the popover version I'm proposing it doesn't work to show loading state on the popover). What we had previously was much cooler in my eyes.

> so that the complexity of implementation of whatever you want to do about the loading could be offloaded off client 

I guess we'd debounce all client operations then apply all at once if this was changed?

bdach commented:
> I guess we'd debounce all client operations then apply all at once if this was changed?

Yes, that's how I'd see this working. I think attempting to make this happen via the current available primitives will result in code that's all of: overly complicated, buggy, and also possibly further negatively impacted by rate limits.

peppy committed:
Move loading inline to individual tags

peppy committed:
Allow multiple in-flight requests

peppy commented:
@bdach i've committed some changes, please see what you think


https://github.com/user-attachments/assets/fccb5872-8fd7-4be2-b17d-2165fab3b645

bdach commented:
@peppy crash happening in the game when viewing this component in its intended location:

```
[runtime] 2025-04-07 11:37:37 [verbose]: Too many unhandled exceptions, crashing out.
Unhandled exception. osu.Framework.Allocation.DependencyNotRegisteredException: The type UserTagControl+AddNewTagUserTag has a dependency on OverlayColourProvider, but the dependency is not registered.
   at osu.Framework.Allocation.ResolvedAttribute.<>c__DisplayClass8_0.<getDependency>b__0(IReadOnlyDependencyContainer dc)
   at osu.Framework.Allocation.ResolvedAttribute.<>c__DisplayClass7_0.<CreateActivator>b__1(Object target, IReadOnlyDependencyContainer dc)
   at osu.Framework.Allocation.DependencyActivator.<Activate>g__activateRecursively|8_0[T](Object obj, IReadOnlyDependencyContainer dependencies, Type currentType)
   at osu.Framework.Allocation.DependencyActivator.Activate[T](T obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Containers.CompositeDrawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Drawable.load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Drawable.Load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies, Boolean isDirectAsyncContext)
```

The results screen doesn't have any overlay colour provider anywhere. Unsure on which colour you want there.

peppy commented:
Let's go with purple for now to match gameplay:

https://github.com/ppy/osu/blob/56a611b47ef44e83b0cd2c21b2e88c91c9f2e6f2/osu.Game/Screens/Play/PlayerLoader.cs#L98-L99

bdach committed:
Cache aquamarine colour provider at results screen

To fix the user tag control.

I would have done it locally to the user tag control, but it was pissing
me off because I wanted the add button to be aquamarine (as it's closer
to the accents the screen is already using), but the popover on open was
for whatever reason purple and I just want consistency where possible.

bdach committed:
Allow tag control popover to attach to bottom too

Looks bad when trying to tag on a replay otherwise.

bdach committed:
Close popovers on hiding the statistics view

Now that the tagging popover is back,
https://github.com/ppy/osu/issues/32630 actually needs fixing.

bdach committed:
Fix crashes in statistics panel test scene

bdach commented:
@peppy Round of changes back at you:

## [Cache aquamarine colour provider at results screen](https://github.com/ppy/osu/pull/32640/commits/23c711d1bc54ac660a94cc0491568f6a969c02d6)

I would have done it locally to the user tag control, but it was pissing me off because I wanted the add button to be aquamarine (as it's closer to the accents the screen is already using), but the popover on open was for whatever reason purple and I just want consistency where possible. See screenshot below on how it looked when I wanted to do it locally:

![Screenshot 2025-04-07 at 14 41 24](https://github.com/user-attachments/assets/ae04cc6b-e237-4df2-99ad-f98dcb41dede)

## [Allow tag control popover to attach to bottom too](https://github.com/ppy/osu/pull/32640/commits/cded1311c8a70ddf9381c92677d08f352fb5c65c)

Looks bad when trying to tag on a replay otherwise:

https://github.com/user-attachments/assets/27cfee26-170d-4b32-bafd-a48c1596edaa

## [Close popovers on hiding the statistics view](https://github.com/ppy/osu/pull/32640/commits/11e48f9f8e805ffbaffeac41fded9cf29f5721d5)

Now that the tagging popover is back, https://github.com/ppy/osu/issues/32630 actually needs fixing:

https://github.com/user-attachments/assets/6ca14395-929c-4ca0-9958-35f7a132f0ed

peppy commented:
Thanks for the fixes. Let's push ahead 👍
